ee/se/modules/380-metallb/hooks/check_requirements_for_upgrade.go

```diff
@@ -111,53 +76,152 @@ func applyIPAddressPoolFilter(obj *unstructured.Unstructured) (go_hook.FilterRes
 	if err != nil {
 		return nil, err
 	}
+	return IPAddressPoolInfo{
+		Name:      ipAddressPool.Name,
+		Namespace: ipAddressPool.Namespace,
+	}, nil
+}
+
+func applyModuleConfigFilter(obj *unstructured.Unstructured) (go_hook.FilterResult, error) {
+	mc := &ModuleConfig{}
+	err := sdk.FromUnstructured(obj, mc)
+	if err != nil {
+		return nil, fmt.Errorf("cannot convert Metallb ModuleConfig: %v", err)
+	}
+	return mc, nil
+}
+
+func applyServiceFilter(obj *unstructured.Unstructured) (go_hook.FilterResult, error) {
+	var service v1.Service
+
+	err := sdk.FromUnstructured(obj, &service)
+	if err != nil {
+		return nil, err
+	}
+
+	if service.Spec.Type != v1.ServiceTypeLoadBalancer {
+		return nil, nil
+	}
+
+	olbsInfo := OrphanedLoadBalancerServiceInfo{
+		Name:       service.Name,
+		Namespace:  service.Namespace,
+		IsOrphaned: true,
+	}
+
+	if service.Spec.LoadBalancerClass != nil {
+		olbsInfo.IsOrphaned = false
+	}
 
-	return ipAddressPool.Name, err
+	if _, ok := service.Annotations["metallb.universe.tf/address-pool"]; ok {
+		olbsInfo.IsOrphaned = false
+	}
+
+	if _, ok := service.Annotations["metallb.universe.tf/ip-allocated-from-pool"]; ok {
+		olbsInfo.IsOrphaned = false
+	}
+
+	return olbsInfo, nil
 }
 
 func checkAllRequirementsForUpgrade(input *go_hook.HookInput) error {
-	ipAddressPoolNamesFromL2A := make([]string, 0, 8)
-	l2AdvertisementSnaps := input.Snapshots["l2advertisements"]
-	for _, l2AdvertisementSnap := range l2AdvertisementSnaps {
-		l2Advertisement := l2AdvertisementSnap.(L2AdvertisementInfo)
-		// Check a namespace
-		if l2Advertisement.Namespace != "d8-metallb" {
-			requirements.SaveValue(metallbConfigurationStatusKey, "NSMismatch")
-			return nil
+	// Disable all alerts
+	for _, alertGroup := range []string{
+		"D8MetallbIpAddressPoolNSMismatch",
+		"D8MetallbL2AdvertisementNSMismatch",
+		"D8MetallbOrphanedLoadBalancerDetected",
+		"D8MetallbL2AdvertisementNodeSelectorsMismatch",
+		"D8MetallbNotOnlyLayer2Pools",
+	} {
+		input.MetricsCollector.Expire(alertGroup)
+	}
+	requirements.SaveValue(metallbConfigurationStatusKey, "OK")
+
+	// Check ModuleConfig version
+	mcSnaps := input.Snapshots["module_config"]
+	if len(mcSnaps) != 1 {
+		return nil
+	}
+	mc, ok := mcSnaps[0].(*ModuleConfig)
+	if !ok || mc.Spec.Version >= 2 {
+		return nil
+	}
+
+	// Are only layer2 pools in the cluster?
+	protocols := make(map[string]bool)
+	l2AdvertisementsCount := len(input.Snapshots["l2_advertisements"])
+	for _, pool := range mc.Spec.Settings.AddressPools {
+		protocols[pool.Protocol] = true
+		if (protocols["layer2"] && protocols["bgp"]) || (protocols["bgp"] && l2AdvertisementsCount > 0) {
```

Comment:
```suggestion
		if (protocols["bgp"] && (protocols["layer2"] || l2AdvertisementsCount > 0)) {
```

Reply:
the map `protocols` is filled in this for loop, maybe better to get it out from loop?

Reply:
I think this kind of algorithm is more efficient. No?